onadata/libs/utils/csv_builder.py

```diff
@@ -622,38 +609,44 @@ def export_to(self, path, dataview=None):
                                          filter_query=self.filter_query)
             if isinstance(cursor, QuerySet):
                 cursor = cursor.iterator()
-            self._update_columns_from_data(cursor)
+
+            self._update_ordered_columns_from_data(cursor)
+
+            data = self._format_for_dataframe(cursor)
 
             columns = list(chain.from_iterable(
                 [[xpath] if cols is None else cols
                  for (xpath, cols) in iteritems(self.ordered_columns)
                  if [c for c in dataview.columns if xpath.startswith(c)]]
             ))
-            cursor = dataview.query_data(dataview, all_data=True,
-                                         filter_query=self.filter_query)
-            if isinstance(cursor, QuerySet):
-                cursor = cursor.iterator()
-            data = self._format_for_dataframe(cursor)
         else:
-            cursor = self._query_data(self.filter_query)
+            try:
+                cursor = self._query_data(self.filter_query)
+            except NoRecordsFoundError:
+                # Set cursor object to an an empty queryset
+                cursor = self.xform.instances.none()
+                # Define export columns using xform schema
+                columns = self.xform.get_headers()
+
+            self._update_ordered_columns_from_data(cursor)
+
             if isinstance(cursor, QuerySet):
                 cursor = cursor.iterator()
-            self._update_columns_from_data(cursor)
+
+            # Unpack xform columns and data
+            data = self._format_for_dataframe(cursor)
 
             columns = list(chain.from_iterable(
                 [[xpath] if cols is None else cols
-                 for (xpath, cols) in iteritems(self.ordered_columns)]))
+                    for (xpath, cols) in iteritems(self.ordered_columns)]))
 
             # add extra columns
             columns += [col for col in self.extra_columns]
+
             for field in self.dd.get_survey_elements_of_type('osm'):
                 columns += OsmData.get_tag_keys(self.xform,
                                                 field.get_abbreviated_xpath(),
                                                 include_prefix=True)
-            cursor = self._query_data(self.filter_query)
-            if isinstance(cursor, QuerySet):
-                cursor = cursor.iterator()
-            data = self._format_for_dataframe(cursor)
```

Comment:
It seems we are loosing the optimization of using the `Queryset.iterator()` which in this case was being used to re-evaluate the records afresh since while generating the columns the data might have been traversed entirely and we need to repeat the same traversal while now sending the data to be written.  Did this have to be moved up?

Reply:
I noticed the data was already converted to an iterator in line 633 - 634. This seemed to be like duplication of code.
I also didnt see any consequence by having `data` definition moved up to line 636. I can move it back to the bottom after the columns has been defined though.